File: ios/StillMoment/Infrastructure/Services/WaveformCacheService.swift
//
//  WaveformCacheService.swift
//  Still Moment
//
//  Infrastructure - Waveform Cache (JSON per meditation)
//

import Foundation
import OSLog

/// Concrete implementation of `WaveformCacheServiceProtocol`.
///
/// Stores one JSON file per meditation under `Application Support/Waveforms/{id}.json`.
/// The directory and `FileManager` are injectable so tests can use a temporary directory.
final class WaveformCacheService: WaveformCacheServiceProtocol {
    // MARK: Lifecycle

    /// - Parameters:
    ///   - fileManager: FileManager instance (defaults to `.default`).
    ///   - directory: Override for the storage directory. When `nil`, resolves to
    ///     `Application Support/Waveforms`.
    init(fileManager: FileManager = .default, directory: URL? = nil) {
        self.fileManager = fileManager
        self.overrideDirectory = directory
    }

    // MARK: Internal

    func load(id: UUID) -> MeditationWaveform? {
        let url = self.fileURL(for: id)
        guard self.fileManager.fileExists(atPath: url.path) else {
            return nil
        }

        do {
            let data = try Data(contentsOf: url)
            return try JSONDecoder().decode(MeditationWaveform.self, from: data)
        } catch {
            Logger.infrastructure.error("Failed to load cached waveform for \(id): \(error.localizedDescription)")
            return nil
        }
    }

    func save(id: UUID, waveform: MeditationWaveform) throws {
        let directory = self.waveformsDirectory()
        if !self.fileManager.fileExists(atPath: directory.path) {
            try self.fileManager.createDirectory(at: directory, withIntermediateDirectories: true)
        }

        let data = try JSONEncoder().encode(waveform)
        try data.write(to: self.fileURL(for: id), options: .atomic)
    }

    func delete(id: UUID) {
        let url = self.fileURL(for: id)
        guard self.fileManager.fileExists(atPath: url.path) else {
            return
        }
        do {
            try self.fileManager.removeItem(at: url)
        } catch {
            Logger.infrastructure.error("Failed to delete cached waveform for \(id): \(error.localizedDescription)")
        }
    }

    // MARK: Private

    private let fileManager: FileManager
    private let overrideDirectory: URL?

    private func waveformsDirectory() -> URL {
        if let overrideDirectory {
            return overrideDirectory
        }
        // Application Support directory is guaranteed to exist on iOS.
        // swiftlint:disable:next force_unwrapping
        let appSupport = self.fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        return appSupport.appendingPathComponent("Waveforms")
    }

    private func fileURL(for id: UUID) -> URL {
        self.waveformsDirectory().appendingPathComponent("\(id.uuidString).json")
    }
}

File: android/app/src/main/kotlin/com/stillmoment/infrastructure/audio/WaveformCacheService.kt
package com.stillmoment.infrastructure.audio

import android.content.Context
import com.stillmoment.domain.models.MeditationWaveform
import com.stillmoment.domain.services.LoggerProtocol
import com.stillmoment.domain.services.WaveformCacheServiceProtocol
import dagger.hilt.android.qualifiers.ApplicationContext
import java.io.File
import java.io.IOException
import javax.inject.Inject
import kotlinx.serialization.SerializationException
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Concrete [WaveformCacheServiceProtocol] (shared-107).
 *
 * Stores one JSON file per meditation under `filesDir/waveforms/{id}.json` (analogous to
 * iOS' `Application Support/Waveforms/`). The directory is injectable so tests can use a
 * temporary directory. Entries are small and never invalidated (non-destructive audio).
 */
class WaveformCacheService(
    private val directory: File,
    private val logger: LoggerProtocol
) : WaveformCacheServiceProtocol {
    @Inject
    constructor(
        @ApplicationContext context: Context,
        logger: LoggerProtocol
    ) : this(File(context.filesDir, WAVEFORMS_DIR), logger)

    override fun load(id: String): MeditationWaveform? {
        val file = fileFor(id)
        if (!file.exists()) {
            return null
        }
        return try {
            Json.decodeFromString<MeditationWaveform>(file.readText())
        } catch (e: SerializationException) {
            logger.e(TAG, "Failed to decode cached waveform for $id", e)
            null
        } catch (e: IOException) {
            logger.e(TAG, "Failed to read cached waveform for $id", e)
            null
        }
    }

    override fun save(id: String, waveform: MeditationWaveform) {
        if (!directory.exists()) {
            directory.mkdirs()
        }
        fileFor(id).writeText(Json.encodeToString(waveform))
    }

    override fun delete(id: String) {
        val file = fileFor(id)
        if (!file.exists()) {
            return
        }
        if (!file.delete()) {
            logger.w(TAG, "Failed to delete cached waveform for $id")
        }
    }

    private fun fileFor(id: String): File = File(directory, "$id.json")

    private companion object {
        const val WAVEFORMS_DIR = "waveforms"
        const val TAG = "WaveformCache"
    }
}
